### PR TITLE
Hyperslab

### PR DESCRIPTION
**Pull request by alkino**
Beginning of #305 
Any thoughts?
You can see example of usage there

### CODE REVIEW

**Comment by alkino** [comment]
We still have to implement `NotB` and `NotA`. I was trying to add a `~` operator to `HyperSlab::Select`.
But I'm not sure to be able to implement everything combination ;)

**Comment by tdegeus** [comment]
Nice!

File: include/highfive/bits/H5Slice_traits_misc.hpp [hunk 1]
```diff
@@ -75,22 +75,9 @@ template <typename Derivate>
...
-
     DataSpace space = slice.getSpace().clone();
-    if (H5Sselect_hyperslab(space.getId(), H5S_SELECT_SET, offset_local.data(),
-                            stride.empty() ? NULL : stride_local.data(),
-                            count_local.data(), NULL) < 0) {
-        HDF5ErrMapper::ToException<DataSpaceException>("Unable to select hyperslap");
-    }
+    HyperSlab(HyperSlab::Select(offset, count, stride)).apply(space);
 
     return Selection(DataSpace(count), space, details::get_dataset(slice));
```
Comment (alkino) [inline include/highfive/bits/H5Slice_traits_misc.hpp:82]: can someone explain me, what means this count?

File: include/highfive/bits/H5Slice_traits.hpp [hunk 2]
```diff
@@ -49,6 +49,122 @@ class ElementSet {
...
+        ret.selects.emplace_back(sel, Op::Xor);
+        return ret;
+    }
+
+    HyperSlab& operator^=(const Select& sel) {
+        selects.emplace_back(sel, Op::Xor);
+        return *this;
+    }
+
+    void apply(const DataSpace& space) {
```
Comment (alkino) [inline include/highfive/bits/H5Slice_traits.hpp:107]: I should make a copy of this DataSpace and return it, right?
From this new DataSpace we can retrieve new dims for the `Selection` class.

**Comment by alkino** [comment]
The problem is that HighFive does not support irregular hyperslab, so this patch is hard to do.

**Comment by ferdonline** [comment]
Progress here looks nice. Do you think we could have tests running?

File: include/highfive/bits/H5Slice_traits_misc.hpp [hunk 3]
```diff
@@ -71,54 +71,38 @@ inline ElementSet::ElementSet(const std::vector<std::vector<std::size_t>>& eleme
     }
 }
 
+template <typename Derivate>
+inline Selection SliceTraits<Derivate>::select(const HyperSlab& hyperslab) const {
+    const auto& slice = static_cast<const Derivate&>(*this);
+    auto space = hyperslab.apply(slice.getSpace());
+
+    return Selection(space, space, details::get_dataset(slice));
```
Comment (alkino) [inline include/highfive/bits/H5Slice_traits_misc.hpp:79]: I think the error is here: before the first arg was `DataSpace(count)` that means nothing here...

File: include/highfive/bits/H5Slice_traits.hpp [hunk 4]
```diff
@@ -49,10 +49,159 @@ class ElementSet {
...
+    DataSpace apply(const DataSpace& space_) const {
+        auto space = space_.clone();
+        for (const auto& sel: selects) {
+            if (sel.op == Op::None) {
+                H5Sselect_none(space.getId());
+            } else {
+                std::vector<hsize_t> offset_local(sel.offset.size());
+                std::vector<hsize_t> count_local(sel.count.size());
+                std::vector<hsize_t> stride_local(sel.stride.size());
+                std::vector<hsize_t> block_local(sel.stride.size());
```
Comment (sergiorg-hpc) [inline include/highfive/bits/H5Slice_traits.hpp:128]: Oups...

**Pull request closed without merging**